shipyard-controller/handler/shipyardcontroller.go

```diff
@@ -241,122 +277,92 @@ func (sc *shipyardController) appendLatestCommitIDToEvent(eventScope models.Even
+	if sequenceExecution == nil {
+		log.Infof("The received %s event with keptn context %s is not accociated with a task that was previously triggered",
 			eventScope.EventType, eventScope.KeptnContext)
-		return nil
+		return ErrSequenceNotFound
 	}
 
-	sc.onSequenceTaskStarted(eventScope.WrappedEvent)
-	return sc.eventRepo.InsertEvent(eventScope.Project, eventScope.WrappedEvent, common.StartedEvent)
-}
-
-func (sc *shipyardController) handleTaskFinished(event models.Event) error {
-	eventScope, err := models.NewEventScope(event)
-	if err != nil {
-		return fmt.Errorf("unable to handle 'task.finished' event: %w", err)
+	if keptnv2.IsStartedEventType(*event.Type) {
+		sc.onSequenceTaskStarted(eventScope.WrappedEvent)
 	}
 
-	if !keptnv2.IsTaskEventType(eventScope.EventType) {
-		return nil
-	}
+	return sc.onTaskProgress(event, *sequenceExecution, eventScope)
+}
 
-	wasTriggered, err := sc.wasTaskTriggered(*eventScope)
+func (sc *shipyardController) onTaskProgress(event models.Event, sequenceExecution models.SequenceExecution, eventScope *models.EventScope) error {
+	taskEvent := models.TaskEvent{
+		EventType: *event.Type,
+		Source:    *event.Source,
+		Result:    eventScope.Result,
+		Status:    eventScope.Status,
+		Time:      event.Time,
+	}
+	if keptnv2.IsFinishedEventType(taskEvent.EventType) {
+		eventData := map[string]interface{}{}
+		err := keptnv2.Decode(event.Data, &eventData)
+		if err != nil {
+			return err
+		}
+		taskEvent.Properties = eventData
+	}
+	updatedSequenceExecution, err := sc.sequenceExecutionRepo.AppendTaskEvent(sequenceExecution, taskEvent)
 	if err != nil {
-		return fmt.Errorf("unable to handle %s event: %w", eventScope.EventType, err)
+		return err
 	}
 
-	if !wasTriggered {
-		log.Infof("The received %s event with keptn context %s is not accociated with a task that was previously triggered",
-			eventScope.EventType, eventScope.KeptnContext)
+	// now check if the number of .started events matches the number of finished events - if yes, that means were done
+	// note: this should also work with multiple replicas because the `AppendTaskEvent` updates the list of events and returns the resulting state
+	// atomically, so ONLY the thread that appended the last event to reach the completion state of the task will get the state required for further proceeding with the task sequence
+	if !updatedSequenceExecution.Status.CurrentTask.IsFinished() {
 		return nil
 	}
 
-	common.LockServiceInStageOfProject(eventScope.Project, eventScope.Stage, eventScope.Service+":taskFinisher")
-	defer common.UnlockServiceInStageOfProject(eventScope.Project, eventScope.Stage, eventScope.Service+":taskFinisher")
+	result, status := updatedSequenceExecution.CompleteCurrentTask()
+
+	eventScope.Result = result
+	eventScope.Status = status
 
-	startedEvents, err := sc.eventRepo.GetStartedEventsForTriggeredID(*eventScope)
+	triggeredEventType, err := keptnv2.ReplaceEventTypeKind(eventScope.EventType, string(common.TriggeredEvent))
```

Comment:
TODO: write todo comment for removing -triggeredEvents collection if we can replace it with sequenceExecution